File: ios/Sasquatsh/ViewModels/EventDetailViewModel.swift
import SwiftUI

@Observable
@MainActor
final class EventDetailViewModel {
    var event: Event?
    var isLoading = false
    var error: String?
    var actionMessage: String?

    private var services: ServiceContainer?

    func configure(services: ServiceContainer) {
        self.services = services
    }

    func loadEvent(id: String) async {
        guard let services else { return }
        isLoading = true
        error = nil
        do {
            event = try await services.events.getEvent(id: id)
        } catch {
            self.error = error.localizedDescription
        }
        isLoading = false
    }

    func register() async {
        guard let services, let event else { return }
        error = nil
        do {
            try await services.events.registerForEvent(eventId: event.id)
            actionMessage = "Registered successfully!"
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func cancelRegistration() async {
        guard let services, let event else { return }
        error = nil
        do {
            try await services.events.cancelRegistration(eventId: event.id)
            actionMessage = "Registration cancelled"
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func claimItem(_ itemId: String) async {
        guard let services, let event else { return }
        error = nil
        do {
            try await services.events.claimItem(itemId: itemId)
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func unclaimItem(_ itemId: String) async {
        guard let services, let event else { return }
        error = nil
        do {
            try await services.events.unclaimItem(itemId: itemId)
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func addItem(name: String, category: String?) async {
        guard let services, let event else { return }
        error = nil
        do {
            let input = CreateEventItemInput(itemName: name, itemCategory: category)
            _ = try await services.events.addItem(eventId: event.id, input: input)
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func addGame(_ input: AddEventGameInput) async {
        guard let services, let event else { return }
        error = nil
        do {
            var gameInput = input
            gameInput.eventId = event.id
            _ = try await services.events.addGame(input: gameInput)
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func removeGame(_ gameId: String) async {
        guard let services, let event else { return }
        error = nil
        do {
            try await services.events.removeGame(gameId: gameId)
            await loadEvent(id: event.id)
        } catch {
            self.error = error.localizedDescription
        }
    }

    func deleteEvent() async -> Bool {
        guard let services, let event else { return false }
        error = nil
        do {
            try await services.events.deleteEvent(id: event.id)
            return true
        } catch {
            self.error = error.localizedDescription
            return false
        }
    }

    func isRegistered(userId: String) -> Bool {
        event?.registrations?.contains { $0.userId == userId } ?? false
    }

    func isHost(userId: String) -> Bool {
        event?.hostUserId == userId
    }
}

File: android/app/src/main/java/com/sasquatsh/app/viewmodels/EventDetailViewModel.kt
package com.sasquatsh.app.viewmodels

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.sasquatsh.app.models.AddEventGameInput
import com.sasquatsh.app.models.CreateEventItemInput
import com.sasquatsh.app.models.Event
import com.sasquatsh.app.services.AuthService
import com.sasquatsh.app.services.EventsService
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import javax.inject.Inject

data class EventDetailUiState(
    val event: Event? = null,
    val isLoading: Boolean = false,
    val error: String? = null,
    val actionMessage: String? = null
)

@HiltViewModel
class EventDetailViewModel @Inject constructor(
    private val eventsService: EventsService,
    private val authService: AuthService,
    savedStateHandle: SavedStateHandle
) : ViewModel() {

    private val _uiState = MutableStateFlow(EventDetailUiState())
    val uiState: StateFlow<EventDetailUiState> = _uiState.asStateFlow()

    private val eventId: String? = savedStateHandle["eventId"]

    init {
        eventId?.let { loadEvent(it) }
    }

    fun loadEvent(id: String) {
        viewModelScope.launch {
            _uiState.update { it.copy(isLoading = true, error = null) }
            try {
                val event = eventsService.getEvent(id)
                _uiState.update { it.copy(event = event, isLoading = false) }
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage, isLoading = false) }
            }
        }
    }

    fun register() {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                eventsService.registerForEvent(event.id)
                _uiState.update { it.copy(actionMessage = "Registered successfully!") }
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun cancelRegistration() {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                eventsService.cancelRegistration(event.id)
                _uiState.update { it.copy(actionMessage = "Registration cancelled") }
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun claimItem(itemId: String) {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                eventsService.claimItem(itemId)
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun unclaimItem(itemId: String) {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                eventsService.unclaimItem(itemId)
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun addItem(name: String, category: String?) {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val input = CreateEventItemInput(itemName = name, itemCategory = category)
                eventsService.addItem(event.id, input)
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun addGame(input: AddEventGameInput) {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                val gameInput = input.copy(eventId = event.id)
                eventsService.addGame(gameInput)
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun removeGame(gameId: String) {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                eventsService.removeGame(gameId)
                loadEvent(event.id)
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun deleteEvent(onSuccess: () -> Unit) {
        val event = _uiState.value.event ?: return
        viewModelScope.launch {
            _uiState.update { it.copy(error = null) }
            try {
                eventsService.deleteEvent(event.id)
                onSuccess()
            } catch (e: Exception) {
                _uiState.update { it.copy(error = e.localizedMessage) }
            }
        }
    }

    fun isRegistered(userId: String): Boolean {
        return _uiState.value.event?.registrations?.any { it.userId == userId } == true
    }

    fun isHost(userId: String): Boolean {
        return _uiState.value.event?.hostUserId == userId
    }

    fun getCurrentUserId(): String? {
        return authService.getCurrentUserId()
    }

    fun clearActionMessage() {
        _uiState.update { it.copy(actionMessage = null) }
    }
}
